Add custom options to editor key incrementation

Samq64 opened the pull request:
Resolves #6555

### Changes

Adds a toggle to use custom values to `editor-number-arrow-keys`.

### Reason for changes

 Having more options could be useful.

### Tests

Tested on Brave.

WorldLanguages commented:
Do we have evidence of an actual user suggesting `100`?

Samq64 commented:
> Do we have evidence of an actual user suggesting `100`?

Just DNin's original post I think.

DNin01 commented:
#3642 would help with the clutter. At the moment, I don't know what to think of so many options one after another.

WorldLanguages commented:
Can we remove 100? :P

DNin01 commented:
> Can we remove 100? :P

I think the only main reason not to include 100 would be due to space constraints.

DNin01 reviewed (approved):
![Single-select overflow](https://github.com/ScratchAddons/ScratchAddons/assets/106490990/ff1aaefe-cdfb-4b90-a4f7-e6e6d2e558a9)
Yeah.

Guess that's just life. We could maybe fix this later.

Samq64 commented:
> Can we remove 100? :P

The popup still has two rows without the 100 option so I don't think it's big deal.

rgantzos commented:
> > Can we remove 100? :P
> 
> The popup still has two rows without the 100 option so I don't think it's big deal.

Even if it doesn't make a difference in terms of the UI, I don't know who would need to change a sprite position by 100, or brightness by 100, or rotation by 100 lol. It's a very large number that it would almost become an inconvenience for many uses.

Samq64 commented:
> Even if it doesn't make a difference in terms of the UI, I don't know who would need to change a sprite position by 100, or brightness by 100, or rotation by 100 lol. It's a very large number that it would almost become an inconvenience for many uses.

I was thinking things like scores but sure I can remove it since it doesn't make much sense outside of variables.

BroJac5246 commented:
> > Even if it doesn't make a difference in terms of the UI, I don't know who would need to change a sprite position by 100, or brightness by 100, or rotation by 100 lol. It's a very large number that it would almost become an inconvenience for many uses.
> 
> I was thinking things like scores but sure I can remove it since it doesn't make much sense outside of variables.

Yes, but if there are any use cases, why get rid of it? It's not like it takes any effort to have it present.

Secret-chest commented:
You can add 2, 4, 8, 16 to fix these :)

rgantzos commented:
There has been a lot of debate about the values. What's wrong with giving full customization and just allowing users to set the value themselves? There can still be a default. At this point the addon settings are limiting the capabilities of the addon.

Samq64 commented:
> There has been a lot of debate about the values. What's wrong with giving full customization and just allowing users to set the value themselves? There can still be a default. At this point the addon settings are limiting the capabilities of the addon.

I think we wanted to avoid making a new setting type since we only support integers right now but I agree that would be better.

BroJac5246 commented:
> > There has been a lot of debate about the values. What's wrong with giving full customization and just allowing users to set the value themselves? There can still be a default. At this point the addon settings are limiting the capabilities of the addon.
> 
> I think we wanted to avoid making a new setting type since we only support integers right now but I agree that would be better.

I don't think we should have arbitrary restrictions just because we don't support decimals in settings.

Secret-chest commented:
Unless we make it customisable, the requests will continue, because we can't satisfy everyone.

Samq64 commented:
I can add float inputs into this PR but I think that needs more discussion and will probably cause the milestone to be bumped.

DNin01 commented:
> I can add float inputs into this PR but I think that needs more discussion and will probably cause the milestone to be bumped.

Let's do that in another PR.

WorldLanguages commented:
> Unless we make it customisable, the requests will continue, because we can't satisfy everyone.

I did not see any requests from the feedback form, literally zero messages about this addon.

Personally I only observed some talk about an `0.5` option in the Discord server this addon was released.

WorldLanguages commented:
> > ...we only support integers right now...
> 
> Wait, why? What's hold it on so we can replace integers in favor of floats or something like that?

Adding more options right now does not mean we will never support non-integer number settings.
I don't really want to turn this PR into a discussion on whether those make sense.

WorldLanguages commented:
Also, let's not forget we do have "string" settings, so we can already support floats, but they won't be validated by the settings page as numbers.

Samq64 commented:
Could 0.125 be removed so the settings fit on one row in the popup? It's also very close to 0.1.

WorldLanguages commented:
> Until we have full customization could 0.125 be removed so the each setting fits on one row in the popup? It's also very close to 0.1.

We could keep the "select" settings intact and add a single toggle for users that want to customize the three exact values through string/float settings.

Secret-chest commented:
> Until we have full customization could 0.125 be removed so the each setting fits on one row in the popup? It's also very close to 0.1.

Noooo! As the suggester of this, I was going to use it not because I like the number, but because it is a power of 2.

Samq64 commented:
I removed the new options and added a custom option instead which currently uses a string setting. If a setting is invalid the block's value changes to NaN which can be undone with Ctrl+Z after unfocusing the input.

BroJac5246 commented:
> > Until we have full customization could 0.125 be removed so the each setting fits on one row in the popup? It's also very close to 0.1.
> 
> Noooo! As the suggester of this, I was going to use it not because I like the number, but because it is a power of 2.

But the whole reason a setting should exist is that people will use it... otherwise why bother?

Also, it's in the thousandths, and that's pretty small.

Secret-chest commented:
> Also, it's in the thousandths, and that's pretty small.

Doesn't matter, it's 1/8.

DNin01 reviewed (approved):
Yeah, that's actually quite a bit better. I can't think of any other ways, so.

WorldLanguages commented:
Screen recording:

![brave_hbmrXMqfdi](https://github.com/ScratchAddons/ScratchAddons/assets/17484114/b2476f09-9251-4403-9bb1-3c2164e56434)

WorldLanguages commented on addons/editor-number-arrow-keys/userscript.js:
```diff
@@ -94,12 +94,21 @@ export default async function ({ addon }) {
-      (e.shiftKey
+    let changeBy = e.code === "ArrowUp" ? 1 : -1;
+    if (addon.settings.get("useCustom")) {
+      // TODO: add proper float settings
+      changeBy *= e.shiftKey
+        ? parseFloat(addon.settings.get("shiftCustom"))
+        : e.altKey
+        ? parseFloat(addon.settings.get("altCustom"))
+        : parseFloat(addon.settings.get("regularCustom"));
```
Yeah, we need to be careful with NaN here. We should early-return or change by zero if the string setting can't be converted into a number.

WorldLanguages commented on addons/editor-number-arrow-keys/addon.json:
```diff
@@ -92,7 +109,23 @@
+    },
+
+    {
+      "name": "Use custom values",
```
We can classify this as a "new" setting with latestUpdate.

WorldLanguages commented:
Note that the fact a number can fit into 8 character doesn't mean it is actually supported, but I decided to draw the line somewhere.

WorldLanguages commented on addons/editor-number-arrow-keys/addon.json:
```diff
@@ -3,6 +3,12 @@
+  "latestUpdate": {
+    "version": "1.35",
+    "isMajor": true,
+    "newSettings": ["useCustom", "regularCustom", "shiftCustom", "altCustom"],
```
I mean, technically all four are new, but...

```suggestion
    "newSettings": ["useCustom"],
```

WorldLanguages commented:
@Secret-chest Hope you're happy now 😂

BroJac5246 commented:
> > Also, it's in the thousandths, and that's pretty small.
> 
> Doesn't matter, it's 1/8.

The same logic applies the other way around, too.

I'm not opposed to this, but only if people will use it. Will they?

Secret-chest commented:
Good.
